gaeguli/fifo-transmit.c

```diff
@@ -613,6 +613,13 @@ _recv_stream (GIOChannel * channel, GIOCondition cond, gpointer user_data)
      * loop indefinitely. */
     guint max_packets = 100;
 
+    if (DEFAULT_ENCODING_METHOD == GAEGULI_ENCODING_METHOD_NVIDIA_TX1) {
```

Comment:
It still feels counter-intuitive why this loop, which should send packets as fast as possible, would actually _increase_ the latency, especially when there seem to be no other critical tasks running in the same main context.

One thing I noticed is that the source pipeline has no queues between the encoder and the filesink, so encoding, MPEG TS muxing and pipe writes are happening all in one thread. It also means when `filesink` is blocked writing into the pipe (which we know **is** a blocking `write()` call), the encoder is blocked as well:

```
#0  0x00007ffff40ec520 in gst_writev_buffers (sink=sink@entry=0x7fffe008ab10 [GstFileSink], fd=25, fdset=fdset@entry=0x0, buffers=buffers@entry=0x7fffe64d4248, num_buffers=num_buffers@entry=1, mem_nums=mem_nums@entry=0x7fffe64d4257 "\001", total_mem_num=1, bytes_written=0x7fffe008adc8, skip=0) at gstelements_private.c:281
#1  0x00007ffff40f2578 in gst_file_sink_render_buffers (sink=sink@entry=0x7fffe008ab10 [GstFileSink], buffers=buffers@entry=0x7fffe64d4248, num_buffers=num_buffers@entry=1, mem_nums=mem_nums@entry=0x7fffe64d4257 "\001", total_mems=1, size=<optimized out>) at gstfilesink.c:660
#2  0x00007ffff40f2d6a in gst_file_sink_render (sink=0x7fffe008ab10 [GstFileSink], buffer=<optimized out>) at gstfilesink.c:825
#3  0x00007ffff4522052 in gst_base_sink_chain_unlocked (basesink=basesink@entry=0x7fffe008ab10 [GstFileSink], obj=obj@entry=0x7fffd422d5a0, is_list=is_list@entry=0, pad=<optimized out>) at gstbasesink.c:3649
#4  0x00007ffff4523110 in gst_base_sink_chain_main (basesink=0x7fffe008ab10 [GstFileSink], pad=<optimized out>, obj=0x7fffd422d5a0, is_list=0) at gstbasesink.c:3775
#5  0x00007ffff77eab0f in gst_pad_chain_data_unchecked (pad=pad@entry=0x7fffe0062a50 [GstPad], type=type@entry=4112, data=data@entry=0x7fffd422d5a0) at gstpad.c:4327
#6  0x00007ffff77ecb71 in gst_pad_push_data (pad=pad@entry=0x7fffe0062800 [GstPad], type=type@entry=4112, data=data@entry=0x7fffd422d5a0) at gstpad.c:4583
#7  0x00007ffff77f3823 in gst_pad_push (pad=0x7fffe0062800 [GstPad], buffer=buffer@entry=0x7fffd422d5a0 [GstBuffer]) at gstpad.c:4702
#8  0x00007fffe65263dd in drain_pending_buffers (parse=parse@entry=0x7fffe0072000 [MpegTSParse2], drain_all=<optimized out>) at mpegtsparse.c:843
#9  0x00007fffe6526ea6 in mpegts_parse_input_done (base=0x7fffe0072000 [MpegTSParse2], buffer=0x0) at mpegtsparse.c:887
#10 0x00007fffe65248bb in mpegts_base_chain (pad=<optimized out>, parent=0x7fffe0072000 [MpegTSParse2], buf=0x7fffd429c360 [GstBuffer]) at mpegtsbase.c:1469
#11 0x00007ffff77eab0f in gst_pad_chain_data_unchecked (pad=pad@entry=0x7fffe00625b0 [GstPad], type=type@entry=4112, data=0x7fffd429c360) at gstpad.c:4327
#12 0x00007ffff77eaeec in gst_pad_chain_list_default (pad=0x7fffe00625b0 [GstPad], parent=<optimized out>, list=0x7fffe0095820 [GstBufferList]) at ../gst/gstbuffer.h:431
#13 0x00007ffff77eaa46 in gst_pad_chain_data_unchecked (pad=pad@entry=0x7fffe00625b0 [GstPad], type=type@entry=4128, data=data@entry=0x7fffe0095820) at gstpad.c:4342
#14 0x00007ffff77ecb71 in gst_pad_push_data (pad=pad@entry=0x7fffe0062360 [GstPad], type=type@entry=4128, data=data@entry=0x7fffe0095820) at gstpad.c:4583
#15 0x00007ffff77f3a63 in gst_pad_push_list (pad=0x7fffe0062360 [GstPad], list=0x7fffe0095820 [GstBufferList]) at gstpad.c:4742
#16 0x00007fffe69e291e in mpegtsmux_push_packets (mux=mux@entry=0x7fffe0069040 [MpegTsMux], force=force@entry=0) at mpegtsmux.c:1671
#17 0x00007fffe69e4955 in mpegtsmux_collected_buffer (pads=pads@entry=0x7fffe006b0d0 [GstCollectPads], data=data@entry=0x7fffe00916f0, buf=0x7fffd4234c60 [GstBuffer], mux=mux@entry=0x7fffe0069040 [MpegTsMux]) at mpegtsmux.c:1508
#18 0x00007ffff453cb9e in gst_collect_pads_default_collected (pads=0x7fffe006b0d0 [GstCollectPads], user_data=<optimized out>) at gstcollectpads.c:1557
#19 0x00007ffff4539d68 in gst_collect_pads_check_collected (pads=pads@entry=0x7fffe006b0d0 [GstCollectPads]) at gstcollectpads.c:1358
#20 0x00007ffff453b696 in gst_collect_pads_chain (pad=0x7fffe0062ca0 [GstPad], parent=<optimized out>, buffer=0x7fffd4234c60 [GstBuffer]) at gstcollectpads.c:2213
#21 0x00007ffff77eab0f in gst_pad_chain_data_unchecked (pad=pad@entry=0x7fffe0062ca0 [GstPad], type=type@entry=4112, data=data@entry=0x7fffd4234c60) at gstpad.c:4327
#22 0x00007ffff77ecb71 in gst_pad_push_data (pad=pad@entry=0x7fffe0062110 [GstPad], type=type@entry=4112, data=data@entry=0x7fffd4234c60) at gstpad.c:4583
#23 0x00007ffff77f3823 in gst_pad_push (pad=0x7fffe0062110 [GstPad], buffer=buffer@entry=0x7fffd4234c60 [GstBuffer]) at gstpad.c:4702
#24 0x00007ffff45116a7 in gst_base_parse_push_frame (parse=parse@entry=0x7fffe0060da0 [GstH264Parse], frame=frame@entry=0x7fffd4004190) at gstbaseparse.c:2580
#25 0x00007ffff451443b in gst_base_parse_handle_and_push_frame (frame=0x7fffd4004190, parse=0x7fffe0060da0 [GstH264Parse]) at gstbaseparse.c:2394
#26 0x00007ffff451443b in gst_base_parse_finish_frame (parse=parse@entry=0x7fffe0060da0 [GstH264Parse], frame=frame@entry=0x7fffd4004190, size=size@entry=8502) at gstbaseparse.c:2738
#27 0x00007fffe65f0f75 in gst_h264_parse_handle_frame (parse=0x7fffe0060da0 [GstH264Parse], frame=0x7fffd4004190, skipsize=0x7fffe64d50d8) at gsth264parse.c:1390
#28 0x00007ffff450c266 in gst_base_parse_handle_buffer (parse=parse@entry=0x7fffe0060da0 [GstH264Parse], buffer=<optimized out>, skip=skip@entry=0x7fffe64d50d8, flushed=flushed@entry=0x7fffe64d50dc) at gstbaseparse.c:2202
#29 0x00007ffff451230e in gst_base_parse_chain (pad=<optimized out>, parent=0x7fffe0060da0 [GstH264Parse], buffer=<optimized out>) at gstbaseparse.c:3287
#30 0x00007ffff77eab0f in gst_pad_chain_data_unchecked (pad=pad@entry=0x7fffe0017db0 [GstPad], type=type@entry=4112, data=data@entry=0x7fffd42a7b40) at gstpad.c:4327
#31 0x00007ffff77ecb71 in gst_pad_push_data (pad=pad@entry=0x7fffe0017b60 [GstPad], type=type@entry=4112, data=data@entry=0x7fffd42a7b40) at gstpad.c:4583
#32 0x00007ffff77f3823 in gst_pad_push (pad=0x7fffe0017b60 [GstPad], buffer=buffer@entry=0x7fffd42a7b40 [GstBuffer]) at gstpad.c:4702
#33 0x00007ffff45c020a in gst_video_encoder_finish_frame (encoder=encoder@entry=0x7fffe005a6f0 [GstX264Enc], frame=0x0, frame@entry=0x7fffd41f4450) at gstvideoencoder.c:2342
#34 0x00007ffff40c323f in gst_x264_enc_encode_frame (encoder=encoder@entry=0x7fffe005a6f0 [GstX264Enc], pic_in=pic_in@entry=0x7fffe64d5510, input_frame=input_frame@entry=0x7fffd41f4450, i_nal=i_nal@entry=0x7fffe64d5508, send=send@entry=1) at gstx264enc.c:2596
#35 0x00007ffff40c7347 in gst_x264_enc_handle_frame (video_enc=0x7fffe005a6f0 [GstX264Enc], frame=0x7fffd41f4450) at gstx264enc.c:2480
#36 0x00007ffff45bedd0 in gst_video_encoder_chain (pad=<optimized out>, parent=0x7fffe005a6f0 [GstX264Enc], buf=<optimized out>) at gstvideoencoder.c:1558
#37 0x00007ffff77eab0f in gst_pad_chain_data_unchecked (pad=pad@entry=0x7fffe0017910 [GstPad], type=type@entry=4112, data=data@entry=0x7fffe0077120) at gstpad.c:4327
...
```
So what if you try putting a queue between the encoder and tsmux?

Reply:
> So what if you try putting a queue between the encoder and tsmux?

 I don't see any enhancement, but that's no harm.
I think it's not bad to have a queue between them.